{{ cookiecutter.formal_name }}/app/src/main/java/org/beeware/android/MainActivity.java

```diff
@@ -55,45 +59,82 @@ private File ensureDir(File dir) throws IOException {
         return dir;
     }
 
-    private Map<String, File> getPythonDirs() throws IOException {
-        Map dirs = new HashMap<String, File>();
+    private Map<String, File> getPythonPaths() throws IOException {
+        Map paths = new HashMap<String, File>();
         File base = new File(getApplicationContext().getFilesDir().getAbsolutePath() + "/python/");
 
         // `stdlib` is used as PYTHONHOME.
         File stdlib = new File(base.getAbsolutePath() + "/stdlib/");
         ensureDir(stdlib);
-        dirs.put("stdlib", stdlib);
+        paths.put("stdlib", stdlib);
+
+        // We cache the stdlib by checking the contents of this file.
+        paths.put("stdlib-last-filename", new File(base.getAbsolutePath() + "/stdlib.last-filename"));
 
         // `rubicon_java` is used to the required Rubicon module.
         File rubicon_java = new File(base.getAbsolutePath() + "/rubicon-java/");
         ensureDir(rubicon_java);
-        dirs.put("rubicon_java", rubicon_java);
+        paths.put("rubicon_java", rubicon_java);
 
-        // Put `user_code` into dirs so we can unpack the assets into it.
-        dirs.put("user_code", new File(base.getAbsolutePath() + "/user_code/"));
+        // Put `user_code` into paths so we can unpack the assets into it.
+        paths.put("user_code", new File(base.getAbsolutePath() + "/user_code/"));
 
         // `app` and `app_packages` store user code and user code dependencies,
         // respectively. These paths exist within the `python/` assets tree.
         File app = new File(base.getAbsolutePath() + "/user_code/app/");
-        dirs.put("app", app);
+        paths.put("app", app);
 
         File app_packages = new File(base.getAbsolutePath() + "/user_code/app_packages/");
-        dirs.put("app_packages", app_packages);
-        return dirs;
+        paths.put("app_packages", app_packages);
+        return paths;
     }
 
-    private void unpackPython(Map<String, File> dirs) throws IOException {
+    private void unpackPython(Map<String, File> paths) throws IOException {
         String myAbi = Build.SUPPORTED_ABIS[0];
-        File pythonHome = dirs.get("stdlib");
-
-        Log.d("unpackPython", "Unpacking Python with ABI " + myAbi + " to " + pythonHome.getAbsolutePath());
-        unzipTo(new ZipInputStream(this.getAssets().open("pythonhome." + myAbi + ".zip")), pythonHome);
-        File rubicon_java = dirs.get("rubicon_java");
+        File pythonHome = paths.get("stdlib");
+
+        // Get list of assets, hoping to find stdlib.
+        String[] assets = this.getAssets().list("");
+        String pythonHomeZipFilename = null;
+        String abiZipSuffix = myAbi + ".zip";
+        for (int i = 0; i < assets.length; i++) {
```

Comment:
I can see why you're doing this, but using a linear probe here makes me a little nervous. Is there any risk that the list that is being iterated over here could be large? Is there any way we could constraint the potential namespace so that the number of files is limited?

Reply:
I agree with you -- the linear probe is a little dangerous in theory! In my very, very limited testing, it wasn't a problem in practice. I didn't time it carefully, but I think that the top level directory shows up first, so it's decently fast.

The best way IMHO to fix this is for the template to place this file into a different directory, like: `app/src/main/assets/pythonhome/{filename}`

I think I'll try to do that, and resubmit this for your review once I do that.